### PR TITLE
 Add filter for Kotlin when-expressions with String

### PR DESCRIPTION
**Pull request by Godin**

### CODE REVIEW

**Comment by marchof** [comment]
@Godin Would prefer to review this after #735 is merged to master.

**Comment by Godin** [comment]
@marchof rebased

Currently diff between `StringSwitchEcjFilter` and `KotlinWhenStringFilter` is

```diff
26,27c26,27
<  * Filters code that is generated by ECJ for a <code>switch</code> statement
<  * with a <code>String</code>.
---
>  * Filters bytecode that Kotlin compiler generates for <code>when</code>
>  * expressions with a <code>String</code>.
29c29
< public final class StringSwitchEcjFilter implements IFilter {
---
> public final class KotlinWhenStringFilter implements IFilter {
46a47
> 			nextIsVar(Opcodes.ALOAD, "s");
73a75,77
> 					// jump to next comparison or default case
> 					nextIs(Opcodes.IFEQ);
> 					final JumpInsnNode jump = (JumpInsnNode) cursor;
75c79
< 					nextIs(Opcodes.IFNE);
---
> 					nextIs(Opcodes.GOTO);
83c87
< 					if (cursor.getNext().getOpcode() == Opcodes.GOTO) {
---
> 					if (jump.label == defaultLabel) {
85,86d88
< 						// jump to default
< 						nextIs(Opcodes.GOTO);
```

😆 but I'd like to propose and would really prefer to work on refactoring separately.

File: org.jacoco.core/src/org/jacoco/core/internal/analysis/filter/KotlinWhenStringFilter.java [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+					}
+				}
+			}
+
+			output.ignore(s.getNext(), cursor);
+			output.replaceBranches(s, replacements);
+		}
+	}
+
+	private static AbstractInsnNode instructionAfterLabel(
```
Comment (marchof) [inline org.jacoco.core/src/org/jacoco/core/internal/analysis/filter/KotlinWhenStringFilter.java]: Idea: Make a variant of `skipNonOpcodes` in base class public:

    private static AbstractInsnNode skipNonOpcodes(AbstractInsnNode start)

**Comment by marchof** [comment]
@Godin Do you want me to merge it as is or do you like to pick-up the `skipNonOpcodes` idea before?

**Comment by Godin** [comment]
@marchof A quick random/thought: `instructionAfterLabel(LabelNode)` compared to `skipNonOpcodes(AbstractInsnNode)` is safer since accepts only `LabelNode`. So let's go without `skipNonOpcodes` - will think about this at a time of thinking about similarity between `StringSwitchEcjFilter` and `KotlinWhenStringFilter`.

**Comment by marchof** [comment]
👍 Let's wait for Travis.

**Pull request merged**